compiler/typecheck/step3_function_and_method_bodies.jou

```diff
@@ -463,34 +474,39 @@ def check_binop(
         or (got_bools and not equals_compare)
         or (got_enums and not equals_compare)
         or (got_pointers and not some_kind_of_compare)
+        or (bitshift and not got_integers)
     ):
-        msg: byte[500]
         snprintf(msg, sizeof(msg), "wrong types: cannot %s %s and %s", do_what, ltype.name, rtype.name)
         fail(location, msg)
 
-    cast_type: Type* = NULL
-    if got_bools:
-        cast_type = boolType
-    if got_integers:
-        cast_type = cast_type_of_integer_binop(ltype, rtype, op)
-    if got_numbers and not got_integers:
-        if ltype == doubleType or rtype == doubleType:
-            cast_type = doubleType
-        else:
-            cast_type = floatType
-    if got_pointers:
-        cast_type = uint_type(64)
-    if got_enums:
-        cast_type = intType
-    assert cast_type != NULL
-
-    do_implicit_cast(fom, lhs, cast_type, Location{}, NULL)
-    do_implicit_cast(fom, rhs, cast_type, Location{}, NULL)
-
-    if some_kind_of_compare:
-        return boolType
+    if op == AstExpressionKind.BitShiftRight:
+        # Special-cased, because right side type does not affect the type of result.
+        do_implicit_cast(fom, rhs, int64Type, Location{}, NULL)
+        return lhs.types.orig_type
     else:
```

Comment:
Let's write this with relying on the early return instead of `else` so that both the diff and resulting code are much more readable.